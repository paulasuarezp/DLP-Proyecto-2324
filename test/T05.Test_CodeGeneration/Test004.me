//*
Test 1 para probar la Fase de Generación de Código

COMPROBACION. Al ejecutar el programa deberá salir:
743,2314

CONSEJO: No limitarse a comprobar el resultado de la ejecución. Se recomienda además abrir el fichero generado usando el depurador GVM y comprobar sus instrucciones.

Para facilitar esta tarea es de gran utilidad que el alumno haya generado directivas '#line' en el fichero de salida.
Para información sobre la función de dicha directiva se debe abrir con GVM el siguiente tutorial: 
Tutorial\3 Uso básico del depurador\4 Fusión Alto Nivel.txt

NOTA: Es posible que el alumno deba adaptar ligeramente la sintaxis del lenguaje si no coincide
con la del lenguaje que le corresponde.
*//

class Test004_HIPOTECA;
		
	global
		/* No hay declaraciones aquí
		
	create main;
		
	/* Features del cálculo
	feature potencia(base: DOUBLE, exponente: INTEGER): DOUBLE is
		local
			total: DOUBLE;
		do
			if exponente = 0 then
				total := 1.0;
			else if exponente < 0 then
					total := 1.0 / potencia(base, 0 - exponente);
				else
					total := base;
					from
					until exponente <= 1 loop
						total := total * base;
						exponente := exponente - 1;
					end
				end
			end
			return total;		
	end
		
	feature letraHipoteca(capital: DOUBLE, meses: INTEGER, interes: DOUBLE): DOUBLE is
		local
			mensual: DOUBLE;
		do
			mensual := interes / 12.0;
			return capital * mensual / (100.0 * (1.0 - potencia(1.0 + mensual/100.0, 0 - meses)));
	end /* letraHipoteca

	feature calcularHipoteca is
		local
			precioViv: DOUBLE;
			mesesHipoteca: INTEGER;
			interes: DOUBLE;
		do			
			/*	Con estos datos deberían salir (redondeando) 743 euros de letra al mes	
			precioViv := 200000.0;
			mesesHipoteca := 30 * 12;
			interes := 2.04;	

			print (letraHipoteca(precioViv, mesesHipoteca, interes));
	end /* calcularHipoteca
		
	feature main is 
		do
			calcularHipoteca();
	end

end

run	main();
