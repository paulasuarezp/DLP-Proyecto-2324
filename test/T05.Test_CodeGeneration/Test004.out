#source "Test004.me"

'Declaraciones globales

call main
halt

'**
'** Features
'**

#line 25
potencia:
#func potencia
#ret float
#param base:float
#param exponente:int

#line 27
enter 4
#local total:float

#line 29
push bp
push 4
add
loadi
pushi 0
eqi
jz else0000

#line 30
push bp
push -4
add
pushf 1.0
storef
jmp finelse0000
else0000:

#line 31
push bp
push 4
add
loadi
pushi 0
lti
jz else0001

#line 32
push bp
push -4
add
pushf 1.0
push bp
push 6
add
loadf
pushi 0
push bp
push 4
add
loadi
subi
call potencia
divf
storef
jmp finelse0001
else0001:

#line 34
push bp
push -4
add
push bp
push 6
add
loadf
storef
' until-from 0002

#line 36
untilcond0002:
push bp
push 4
add
loadi
pushi 1
lei
jnz untilend0002

#line 36

#line 37
push bp
push -4
add
push bp
push -4
add
loadf
push bp
push 6
add
loadf
mulf
storef

#line 38
push bp
push 4
add
push bp
push 4
add
loadi
pushi 1
subi
storei
jmp untilcond0002
untilend0002:
finelse0001:
finelse0000:

#line 42
push bp
push -4
add
loadf
ret 4, 4, 6

#line 45
letraHipoteca:
#func letraHipoteca
#ret float
#param capital:float
#param meses:int
#param interes:float

#line 47
enter 4
#local mensual:float

#line 49
push bp
push -4
add
push bp
push 4
add
loadf
pushf 12.0
divf
storef

#line 50
push bp
push 10
add
loadf
push bp
push -4
add
loadf
mulf
pushf 100.0
pushf 1.0
pushf 1.0
push bp
push -4
add
loadf
pushf 100.0
divf
addf
pushi 0
push bp
push 8
add
loadi
subi
call potencia
subf
mulf
divf
ret 4, 4, 10

#line 53
calcularHipoteca:
#func calcularHipoteca
#ret void

#line 57
enter 10
#local precioViv:float
#local mesesHipoteca:int
#local interes:float

#line 60
push bp
push -4
add
pushf 200000.0
storef

#line 61
push bp
push -6
add
pushi 30
pushi 12
muli
storei

#line 62
push bp
push -10
add
pushf 2.04
storef

#line 64
push bp
push -4
add
loadf
push bp
push -6
add
loadi
push bp
push -10
add
loadf
call letraHipoteca
outf

#line 65
ret 0, 10, 0

'**
'** Main
'**

#line 70
main:
#func main
#ret void

#line 70
call calcularHipoteca

#line 71
ret 0, 0, 0
